lib/Sema/CSDiag.cpp

```diff
@@ -8403,3 +8412,83 @@ bool swift::diagnoseBaseUnwrapForMemberAccess(Expr *baseExpr, Type baseType,
   }
   return true;
 }
+
+void swift::diagnoseIllegalEphemeralConversion(
+    ASTContext &ctx, const Expr *argExpr, Type argType, Type paramType,
+    const ValueDecl *callee, AnyFunctionType *fnType, const Expr *anchor,
+    bool downgradeToWarning) {
+  auto &diags = ctx.Diags;
+
+  auto emitArgumentConversionNote = [&]() {
+    diags.diagnose(argExpr->getLoc(),
+                   diag::ephemeral_pointer_argument_conversion_note,
+                   argType->getWithoutSpecifierType(), paramType)
+        .highlight(argExpr->getSourceRange());
+  };
+
+  // Emit a specialised diagnostic for
+  // Unsafe[Mutable][Raw]Pointer.init([mutating]:) &
+  // Unsafe[Mutable][Raw]BufferPointer.init(start:count:).
+  auto diagnosePointerInit = [&]() -> bool {
+    auto *constructor = dyn_cast_or_null<ConstructorDecl>(callee);
+    if (!constructor)
+      return false;
+
+    auto instanceTy = fnType->getResult();
+
+    // Strip off a level of optionality if we have a failable initializer.
+    switch (constructor->getFailability()) {
+    case OTK_Optional:
+    case OTK_ImplicitlyUnwrappedOptional:
+      instanceTy = instanceTy->getOptionalObjectType();
+      break;
+    case OTK_None:
+      break;
+    }
+
+    auto isPointerType = [](Type ty, const ASTContext &ctx) -> bool {
+      if (auto nominalTy = ty->getAs<NominalType>())
+        if (nominalTy->getDecl() == ctx.getOpaquePointerDecl())
+          return true;
+      return (bool)ty->getAnyPointerElementType();
+    };
+
+    // This must stay in sync with diag::cannot_construct_dangling_pointer.
+    enum ConstructorKind {
+      CK_Pointer = 0,
+      CK_BufferPointer,
+    };
+
+    ConstructorKind constructorKind;
+    auto parameterCount = constructor->getParameters()->size();
+    if (isPointerType(instanceTy, ctx) && parameterCount == 1) {
+      constructorKind = CK_Pointer;
+    } else if (instanceTy->getAnyBufferPointerElementType() &&
+               parameterCount == 2) {
+      constructorKind = CK_BufferPointer;
+    } else {
+      return false;
+    }
+
+    auto diagID = downgradeToWarning
+                      ? diag::cannot_construct_dangling_pointer_warning
+                      : diag::cannot_construct_dangling_pointer;
+
+    diags.diagnose(anchor->getLoc(), diagID, instanceTy, constructorKind)
+        .highlight(anchor->getSourceRange());
```

Comment:
Is the anchor expression always the correct expression? What if the construction is nested inside a larger expression?

Reply:
As I understand it, unsimplified argument locators for an apply/subscript are always anchored on the apply/subscript expr that they correspond to.

This is how they're generated in CSApply and CSGen at least:

https://github.com/apple/swift/blob/af1391400940b41ee50d4766405aa538133b279a/lib/Sema/CSGen.cpp#L2561-L2564

https://github.com/apple/swift/blob/af1391400940b41ee50d4766405aa538133b279a/lib/Sema/CSApply.cpp#L3085-L3089